### ISSUE TITLE
temporary directory fixture

### ISSUE THREAD

**Issue by mbway** [la10736/rstest#144]
pytest provides a very useful fixture: [tmp_path](https://docs.pytest.org/en/7.1.x/reference/reference.html?highlight=tmp_path#tmp-path) which tests can use to be given a temporary directory unique to that test.
```python
import pytest
from pathlib import Path

def test_something(tmp_path: Path):
    print(tmp_path)  # e.g. /tmp/pytest-of-user/pytest-1/test_something/
```

The fixture manages temporary directories in `/tmp/pytest-of-<username>/pytest-<counter>/<name-of-test>/`. Each time pytest is run the counter is incremented and results older than a threshold are removed.

Some advantages of using a fixture like this over creating a temporary directory yourself:
- ease of use
- the temporary location is easy to find after the test has run
    - the test name is used as part of the path
- the fixture can be used as arguments to other fixtures so all the data for the test can be written to the same place
```python
@pytest.fixture()
def input_data(tmp_path: Path) -> Path:
    return tmp_path / "input_data"  # /tmp/pytest-of-user/pytest-1/test_thing/input_data

@pytest.fixture()
def expected_output(tmp_path: Path) -> Path:
    return tmp_path / "expected_output"  # /tmp/pytest-of-user/pytest-1/test_thing/expected_output

def test_thing(tmp_path: Path, input_data: Path, expected_output: Path):
    output_path = tmp_path / "output"
    do_something(input_data, output_path)
    compare_results(output_path, expected_output)

```

Would it be possible to create a fixture like this as an rstest fixture? I can't see any way that the test name could be passed to a fixture.

**Comment by la10736** [comment]
Yes, it is possible and non so hard. We can split it in some sub fixture like `test_name` and `tmp_path`.

For the name you can use just `thread::current().name().unwrap().to_string()` (the running thread have the same name).

I can plan in the next moth.. unfortunately I haven't any time in these days.

**Comment by TimeEngineer** [comment]
Hi,
We also use tmp_path in our project to mimic pytest and to make it easier to delete the files generated by the tests.

Currently, we have implemented a drop that will clean the tmp_path folder at the end of the test but we end up with the problem that when a fixture calls tmp_path, we must be able to keep it alive so that it is cleaned at the end, this requires implementing a new structure for each fixture that needs tmp_path.

Do you have another solution in mind to solve the problem?

**Comment by svenstaro** [comment]
Just stumbled upon this. I just use:
```rust
/// Test fixture which creates a temporary directory
#[fixture]
pub fn tmpdir() -> TempDir {
    assert_fs::TempDir::new().expect("Couldn't create a temp dir for tests")
}
```

**Comment by mbway** [comment]
right, but the `tmp_path` fixture gives you more than that. When running a large suite of tests you can easily find the directory that corresponds to a failing test and once you have run the test suite a number of times, old runs are cleared etc.

**Comment by svenstaro** [comment]
Ah I see. That would be quite handy indeed!

**Comment by la10736** [comment]
I'm using a crate of mine to test retest. You can obtain quite the same
result that you have with the pytest's fixture.
https://crates.io/crates/temp_testdir/0.2.3

Il sab 25 mar 2023, 00:03 Sven-Hendrik Haase ***@***.***> ha
scritto:
